src/rover/camera.hpp

```diff
@@ -0,0 +1,225 @@
+
+    Parameters:
+        session: The session struct used to maintain session state.
+        device_filepath: The file path of the camera device (/dev/video<n>, where n = 0, 1, ...).
+        width, height: The capture resolution.
+    
+    Returns:
+        Error::OK on success, and a suitable error on failure:
+            Error::OPEN: Failed to open the camera device file.
+            Error::QUERY_CAPABILITIES: Failed to retrieve camera device capabilities.
+            Error::NO_VIDEO_CAPTURE: The device does not support video capture.
+            Error::NO_STREAMING: The device does not support video streaming.
+            Error::QUERY_FORMAT: Failed to retrieve device pixel format and resolution.
+            Error::SET_FORMAT: Failed to offer our desired format to the device.
+            Error::UNSUPPORTED_FORMAT: The device does not support our format.
+            Error::UNSUPPORTED_RESOLUTION: The device does not support the supplied resolution.
+            Error::REQUEST_BUFFERS: Failed to request the ability to use our buffers.
+            Error::LINK_BUFFERS: Failed to link our buffers to the device.
+*/
+Error open(CaptureSession* session, const char* device_filepath, size_t width, size_t height);
+
+/*
+    Starts the capture session. The device will begin to offer frames.
+
+    Parameters:
+        session: The capture session. This must first be initialized with `open`.
+    
+    Returns:
+        Error::OK on success, and a suitable error on failure:
+            Error::START_STREAM: Failed to start the capture stream.
+*/
+Error start(CaptureSession* session);
+
+/*
+    Grabs the next frame from the camera device.
+
+    The frame buffer MUST be returned after it is processed by calling `return_buffer`.
+
+    Parameters:
+        session: The capture session. `start` must be called prior to this function.
+    
+    Return Parameters:
+        out_frame: The frame data (on success).
+        out_frame_size: The size of the captured frame (on success).
+
+    Returns:
+        Error::OK when a frame was read.
+        Otherwise, a suitable error is returned:
+            Error::SELECT: Failed while waiting to read from the device.
+            Error::READ_FRAME: Failed to read raw frame data.
+*/
+Error grab_frame(CaptureSession* session, uint8_t** out_frame, size_t* out_frame_size);
+
+/*
+    Returns the frame buffer so that V4L can reuse it. MUST be called after
+    `grab_frame`.
+
+    Parameters:
+        session: The capture session.
+        buffer: A buffer returned by `grab_frame`.
+    
+    Returns:
+        Error::OK on success and a suitable error on failure:
+            Error::PREPARE_BUFFER: Failed to prepare the buffer for another read.
+*/
+Error return_buffer(CaptureSession* session, uint8_t* buffer);
+
+/*
+    Stops capture and closes the device.
+
+    Parameters:
+        session: The capture session.
+*/
+void close(CaptureSession* session);
+
+} // namespace camera
+
+#endif
```

Comment:
Can't comment below, but why did you check in a binary file? Was that intentional? (camera.obj)

Reply:
Nah, I've taken it out